services/sendgrid/v2/emailService.go

```diff
@@ -112,5 +112,29 @@ func (s *sendgridEmailService) SendEmailVerificationEmail(ctx context.Context, u
 }
 
 func (s *sendgridEmailService) SendPasswordResetEmail(ctx context.Context, user entities.User, passwordResetResources common.UniformResourceIdentifiers) error {
-	panic("not implemented")
+	emailToken, err := s.authorizer.CreateServiceToken(ctx, user.ID,
```

Comment:
We're creating a service token for each password reset email, at the moment we don't remove it from the `token` collection once the password is changed. This means that the same token could be used multiple times to reset an email.

Do you think we should create another issue to invalidate the service token once it has been used to reset the password?

Reply:
You're right. I should implement invalidating the service token in this PR since it's just something  I forgot to do.

Reply:
Right, so I implemented the functionality to invalidate the token after the password is reset. However I realized that by doing this I made the operation `SetPassword` basically unusable with service tokens (same goes for the `VerifyEmail` operation) as calling it would result in the token being invalidated. This is probably fine for now since I don't see any reason why we'd want to change the password of the user from anywhere outside of `hs_auth` but we should still fix it at some point. I've labeled this as a bug and added a TODO: https://github.com/unicsmcr/hs_auth/issues/121

Reply:
Good call, I think the only time we'd use a service token for `SetPassword` is if we had a form on the user profile to change their password (which I don't think we'd need anyway). It's a good idea to fix the bug in anycase 👍

Reply:
I think it'd be beneficial to fix this bug just because we'd need to design a more elegant solution for single use email tokens. The way we handle this logic right now is definitely not pretty. I've added a description for a solution that I think would work quite well to the description of the issue.